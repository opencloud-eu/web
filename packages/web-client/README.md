# web-client

The `web-client` is a standalone package that allows you to interact with the [OpenCloud](https://github.com/opencloud-eu/opencloud/) APIs via TypeScript. It provides an abstraction layer between the server and a (web-) application that converts API data into objects with helpful types and utilities. This abstraction ensures that users of the APIs don't need in-depth knowledge about them, such as required methods or returned status codes.

The supported APIs are:

- Graph (drives, sharing, user & group management)
- OCS (capabilities & url signing)
- WebDAV (file operations)

## Installation

Depending on your package manager, run one of the following commands:

```
$ npm install @opencloud-eu/web-client

$ pnpm add @opencloud-eu/web-client

$ yarn add @opencloud-eu/web-client
```

## Usage

### Graph

The graph client needs to be instantiated with a base URI corresponding to your OpenCloud deployment and an axios instance. The axios instance is being used for all requests, which means it needs to include all relevant headers either statically or via interceptor.

```
import axios from axios
import { graph } from '@opencloud-eu/web-client'

const accessToken = 'some_access_token'
const baseURI = 'some_base_uri'

const axiosClient = axios.create({
	headers: { Authorization: accessToken }
})

const graphClient = graph(baseURI, axiosClient)
```

The following example demonstrates how to retrieve all spaces accessible to the user. A `SpaceResource` can then be used to e.g. fetch files and folders (see webdav example down below).

```
const mySpaces = await graphClient.drives.listMyDrives()
```

### OCS

The ocs client needs to be instantiated with a base URI corresponding to your OpenCloud deployment and an axios instance. The axios instance is being used for all requests, which means it needs to include all relevant headers either statically or via interceptor.

```
import axios from axios
import { ocs } from '@opencloud-eu/web-client'

const accessToken = 'some_access_token'
const baseURI = 'some_base_uri'

const axiosClient = axios.create({
	headers: { Authorization: accessToken }
})

const ocsClient = ocs(baseURI, axiosClient)
```

The following examples demonstrate how to fetch capabilities and sign URLs.

```
const capabilities = await ocsClient.getCapabilities()

const signedUrl = await ocsClient.signUrl('some_url_to_sign', 'your_username')
```

### WebDav

The webdav client needs to be instantiated with a base URI corresponding to your OpenCloud deployment. You can also pass a header callback which will be called with every dav request.

```
import { webdav } from '@opencloud-eu/web-client'

const accessToken = 'some_access_token'
const baseURI = 'some_base_uri'

const webDavClient = webdav(
	baseURI,
	() => ({ Authorization: accessToken })
)
```

The following example demonstrates how to list all resources of a given `SpaceResource` (see above how to fetch space resources).

```
const { resource, children } = await webDavClient.listFiles(spaceResource)
```
