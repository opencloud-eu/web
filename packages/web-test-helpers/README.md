# web-test-helpers

This packages provides utilities for unit testing within the OpenCloud app ecosystem.

## Installation

Depending on your package manager, run one of the following commands:

```
$ npm install @opencloud-eu/web-test-helpers --save-dev

$ pnpm add -D @opencloud-eu/web-test-helpers

$ yarn add @opencloud-eu/web-test-helpers --dev
```

Make sure that you have `vue`, `@vue/test-utils`, and `@opencloud-eu/web-pkg` as dev dependencies of your app since those are peer dependencies of this package.
