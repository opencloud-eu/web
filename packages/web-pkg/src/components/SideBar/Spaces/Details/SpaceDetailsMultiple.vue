<template>
  <div id="oc-spaces-details-multiple-sidebar">
    <div class="spaces-preview oc-mb-l">
      <div class="spaces-preview-body">
        <oc-icon class="preview-icon" size="xxlarge" name="layout-grid" />
        <p class="preview-text" v-text="selectedSpacesString" />
      </div>
    </div>
    <oc-definition-list :aria-label="detailsTableLabel" :items="items" />
  </div>
</template>
<script lang="ts">
import { formatFileSize } from '../../../../helpers'
import { computed, defineComponent, PropType, unref } from 'vue'
import { SpaceResource } from '@opencloud-eu/web-client'
import { useGettext } from 'vue3-gettext'

export default defineComponent({
  name: 'SpaceDetailsMultiple',
  props: {
    selectedSpaces: {
      type: Array as PropType<Array<SpaceResource>>,
      required: true
    }
  },
  setup(props) {
    const language = useGettext()
    const { $gettext, $ngettext } = language
    const totalSelectedSpaceQuotaTotal = computed(() => {
      let total = 0
      props.selectedSpaces.forEach((space) => {
        total += space.spaceQuota.total
      })
      return formatFileSize(total, language.current)
    })
    const totalSelectedSpaceQuotaRemaining = computed(() => {
      let remaining = 0
      props.selectedSpaces.forEach((space) => {
        if (space.disabled) {
          return
        }
        remaining += space.spaceQuota.remaining
      })
      return formatFileSize(remaining, language.current)
    })
    const totalSelectedSpaceQuotaUsed = computed(() => {
      let used = 0
      props.selectedSpaces.forEach((space) => {
        if (space.disabled) {
          return
        }
        used += space.spaceQuota.used
      })
      return formatFileSize(used, language.current)
    })
    const totalEnabledSpaces = computed(() => {
      return props.selectedSpaces.filter((s) => !s.disabled).length
    })
    const totalDisabledSpaces = computed(() => {
      return props.selectedSpaces.filter((s) => s.disabled).length
    })
    const detailsTableLabel = computed(() => {
      return $gettext('Overview of the information about the selected spaces')
    })
    const selectedSpacesString = computed(() => {
      return $ngettext(
        '%{ itemCount } space selected',
        '%{ itemCount } spaces selected',
        props.selectedSpaces.length,
        {
          itemCount: props.selectedSpaces.length.toString()
        }
      )
    })

    const items = [
      { term: $gettext('Total quota:'), definition: unref(totalSelectedSpaceQuotaTotal) },
      { term: $gettext('Remaining quota:'), definition: unref(totalSelectedSpaceQuotaRemaining) },
      { term: $gettext('Used quota:'), definition: unref(totalSelectedSpaceQuotaUsed) },
      { term: $gettext('Enabled:'), definition: unref(totalEnabledSpaces) },
      { term: $gettext('Disabled:'), definition: unref(totalDisabledSpaces) }
    ]

    return {
      detailsTableLabel,
      items,
      selectedSpacesString
    }
  }
})
</script>
<style lang="scss" scoped>
#oc-spaces-details-multiple-sidebar {
  background-color: var(--oc-role-surface-container);
  border-radius: 5px;
  padding: var(--oc-space-medium);
}

.details-list {
  margin: 0;
}

.spaces-preview {
  text-align: center;
  border-radius: 5px;
}
</style>
