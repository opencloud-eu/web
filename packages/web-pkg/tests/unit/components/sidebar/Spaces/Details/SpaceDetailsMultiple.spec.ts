import { SpaceResource } from '@opencloud-eu/web-client'
import SpaceDetailsMultiple from '../../../../../../src/components/SideBar/Spaces/Details/SpaceDetailsMultiple.vue'
import { defaultPlugins, shallowMount } from '@opencloud-eu/web-test-helpers'

const spaceMock = {
  type: 'space',
  name: ' space',
  id: '1',
  mdate: 'Wed, 21 Oct 2015 07:28:00 GMT',
  spaceQuota: {
    used: 100,
    total: 1000,
    remaining: 900
  }
} as unknown as SpaceResource

describe('Multiple Details SideBar Panel', () => {
  it('displays the details side panel', () => {
    const { wrapper } = createWrapper(spaceMock)
    expect(wrapper.html()).toMatchSnapshot()
  })
})

function createWrapper(spaceResource: SpaceResource) {
  return {
    wrapper: shallowMount(SpaceDetailsMultiple, {
      global: {
        plugins: [...defaultPlugins()]
      },
      props: {
        selectedSpaces: [spaceResource]
      }
    })
  }
}
