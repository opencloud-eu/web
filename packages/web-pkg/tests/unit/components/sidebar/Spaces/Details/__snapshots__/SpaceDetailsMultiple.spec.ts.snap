// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Multiple Details SideBar Panel > displays the details side panel 1`] = `
"<div data-v-cd05ba5e="" id="oc-spaces-details-multiple-sidebar">
  <div data-v-cd05ba5e="" class="spaces-preview oc-mb-l">
    <div data-v-cd05ba5e="" class="spaces-preview-body">
      <oc-icon-stub data-v-cd05ba5e="" accessiblelabel="" color="" filltype="fill" name="layout-grid" size="xxlarge" type="span" class="preview-icon"></oc-icon-stub>
      <p data-v-cd05ba5e="" class="preview-text">1 space selected</p>
    </div>
  </div>
  <oc-definition-list-stub data-v-cd05ba5e="" items="[object Object],[object Object],[object Object],[object Object],[object Object]" aria-label="Overview of the information about the selected spaces"></oc-definition-list-stub>
</div>"
`;
